File: plugins/maestro/skills/maestro/scripts/maestro_config.py
#!/usr/bin/env python3
"""Read or update Maestro's saved skill-bundle-folder."""

from __future__ import annotations

import argparse
import json
import pathlib
from typing import Dict, Optional

try:
    import maestro_fetch as fetch
except ImportError as exc:  # pragma: no cover
    raise SystemExit(f"cannot import maestro_fetch.py from script directory: {exc}")


def result(status: str, skill_file: pathlib.Path, bundle_root: Optional[pathlib.Path] = None, message: Optional[str] = None) -> Dict[str, object]:
    payload: Dict[str, object] = {
        "status": status,
        "skill_file": str(skill_file),
        "config_file": str(fetch.config_file(skill_file)),
    }
    if bundle_root is not None:
        payload["bundle_root"] = str(bundle_root)
        payload["exists"] = bundle_root.is_dir()
    if message:
        payload["message"] = message
    return payload


def main() -> int:
    parser = argparse.ArgumentParser(description="Read or update Maestro's saved skill-bundle-folder.")
    parser.add_argument("command", choices=("get", "set"), help="Use get to read the saved path, set to replace it.")
    parser.add_argument("bundle_root", nargs="?", help="New skill-bundle-folder path for set.")
    parser.add_argument(
        "--skill-file",
        default=str(pathlib.Path(__file__).resolve().parents[1] / "SKILL.md"),
        help="Maestro SKILL.md path. Defaults to this skill's SKILL.md.",
    )
    parser.add_argument("--text", action="store_true", help="Print compact text instead of JSON.")
    args = parser.parse_args()

    skill_file = pathlib.Path(args.skill_file).expanduser().resolve()
    if not skill_file.is_file():
        raise SystemExit(f"skill file not readable: {skill_file}")

    if args.command == "get":
        bundle_root = fetch.load_saved_bundle_root(skill_file)
        if bundle_root is None:
            payload = result("missing", skill_file, message="bundle root not configured")
        else:
            payload = result("configured", skill_file, bundle_root=bundle_root)
    else:
        if not args.bundle_root:
            raise SystemExit("bundle_root required for set")
        bundle_root = pathlib.Path(args.bundle_root).expanduser().resolve()
        if not bundle_root.is_dir():
            raise SystemExit(f"bundle root not readable: {bundle_root}")
        fetch.save_bundle_root(skill_file, bundle_root)
        payload = result("configured", skill_file, bundle_root=bundle_root)

    if args.text:
        print(payload["status"])
        if "bundle_root" in payload:
            print(f"bundle_root: {payload['bundle_root']}")
        if "message" in payload:
            print(payload["message"])
    else:
        print(json.dumps(payload, indent=2, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: maestro/scripts/maestro_config.js
#!/usr/bin/env node
"use strict";
/**
 * Read or update Maestro's saved skill-bundle-folder.
 * Node.js port of maestro_config.py.
 */

const fs   = require("fs");
const path = require("path");
const fetch = require("./maestro_fetch.js");

function result(status, skillFile, bundleRoot, message) {
  const payload = {
    status,
    skill_file: skillFile,
    config_file: fetch.configFile(skillFile),
  };
  if (bundleRoot != null) {
    payload.bundle_root = bundleRoot;
    payload.exists = fs.existsSync(bundleRoot) && fs.statSync(bundleRoot).isDirectory();
  }
  if (message) {
    payload.message = message;
  }
  return payload;
}

function main() {
  const args = process.argv.slice(2);
  let command      = null;
  let bundleRootArg = null;
  let skillFileArg  = path.resolve(path.join(__dirname, "..", "SKILL.md"));
  let textMode      = false;

  for (let i = 0; i < args.length; i++) {
    if (args[i] === "--skill-file" && args[i + 1]) { skillFileArg = args[++i]; }
    else if (args[i] === "--text") { textMode = true; }
    else if (!args[i].startsWith("-") && !command) { command = args[i]; }
    else if (!args[i].startsWith("-")) { bundleRootArg = args[i]; }
  }

  if (command !== "get" && command !== "set") {
    console.error("Usage: node maestro_config.js <get|set> [bundle_root] [--skill-file <path>] [--text]");
    process.exit(1);
  }

  const skillFile = path.resolve(skillFileArg);
  if (!fs.existsSync(skillFile)) {
    console.error(`skill file not readable: ${skillFile}`);
    process.exit(1);
  }

  let payload;
  if (command === "get") {
    const bundleRoot = fetch.loadSavedBundleRoot(skillFile);
    if (!bundleRoot) {
      payload = result("missing", skillFile, null, "bundle root not configured");
    } else {
      payload = result("configured", skillFile, bundleRoot);
    }
  } else {
    if (!bundleRootArg) {
      console.error("bundle_root required for set");
      process.exit(1);
    }
    const bundleRoot = path.resolve(bundleRootArg);
    try {
      if (!fs.statSync(bundleRoot).isDirectory()) throw new Error();
    } catch {
      console.error(`bundle root not readable: ${bundleRoot}`);
      process.exit(1);
    }
    fetch.saveBundleRoot(skillFile, bundleRoot);
    payload = result("configured", skillFile, bundleRoot);
  }

  if (textMode) {
    console.log(payload.status);
    if (payload.bundle_root) console.log(`bundle_root: ${payload.bundle_root}`);
    if (payload.message) console.log(payload.message);
  } else {
    console.log(JSON.stringify(payload, null, 2));
  }
}

if (require.main === module) {
  main();
}
